### ISSUE TITLE
Cannot use include in task handlers (AttributeError)

### ISSUE THREAD

**Issue by pstch** [ansible/ansible#15915]
##### ISSUE TYPE
- Bug Report
##### ANSIBLE VERSION

**Affected :**

```
ansible 2.2.0
  config file = /home/pistache/.ansible.cfg
  configured module search path = Default w/o overrides
```

```
ansible 2.1.0.0
  config file = /home/pistache/.ansible.cfg
  configured module search path = Default w/o overrides

```

**Not affected :** Ansible 2.0.2
##### CONFIGURATION

None
##### OS / ENVIRONMENT

Debian Unstable (on my workstation, bug was also reproduced with Ansible 2.1 on Jessie)
##### SUMMARY

I cannot use `include` in a task handler with Ansible 2.1 or 2.2 (`stable-2.1` and `devel`), the same code that was working with Ansible 2.0 now raises an AttributeError.

It seems like the task handler code was modified in such a way that it does not allow `include` anymore.
##### STEPS TO REPRODUCE

Create a role, and use `include` in a task handler.

`handlers/main.yml`

```
- name: Test
  include: test.yml
```

`handlers/test.yml`

```
- debug: msg="Test handler notified."
```

`tasks/main.yml`

```
- name: Test task
  debug: msg="Notifying test handler"
  changed_when: yes
  notify:
    - Test
```
##### EXPECTED RESULTS

No errors, should print out "Test handler notified."
##### ACTUAL RESULTS

```
ERROR! Unexpected Exception: 'TaskInclude' object has no attribute 'has_triggered'
the full traceback was:

Traceback (most recent call last):
  File "/usr/local/bin/ansible-playbook", line 92, in <module>
    exit_code = cli.run()
  File "/usr/local/lib/python2.7/dist-packages/ansible/cli/playbook.py", line 154, in run
    results = pbex.run()
  File "/usr/local/lib/python2.7/dist-packages/ansible/executor/playbook_executor.py", line 142, in run
    result = self._tqm.run(play=play)
  File "/usr/local/lib/python2.7/dist-packages/ansible/executor/task_queue_manager.py", line 253, in run
    play_return = strategy.run(iterator, play_context)
  File "/usr/local/lib/python2.7/dist-packages/ansible/plugins/strategy/linear.py", line 215, in run
    self._execute_meta(task, play_context, iterator)
  File "/usr/local/lib/python2.7/dist-packages/ansible/plugins/strategy/__init__.py", line 707, in _execute_meta
    self.run_handlers(iterator, play_context)
  File "/usr/local/lib/python2.7/dist-packages/ansible/plugins/strategy/__init__.py", line 589, in run_handlers
    result = self._do_handler_run(handler, handler_name, iterator=iterator, play_context=play_context)
  File "/usr/local/lib/python2.7/dist-packages/ansible/plugins/strategy/__init__.py", line 621, in _do_handler_run
    if not handler.has_triggered(host) and (host.name not in self._tqm._failed_hosts or play_context.force_handlers):
AttributeError: 'TaskInclude' object has no attribute 'has_triggered'

```

**Comment by Schnitzel** [comment]
had the same issue and found the reason, see https://github.com/ansible/ansible/pull/15958

**Comment by jimi-c** [comment]
# Closing This Ticket

@pstch I've merged in @Schnitzel's PR above, so this should be resolved. This was also merged into stable-2.1 and will be included in the next release.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!

**Comment by xenithorb** [comment]
still seeing this is an unpatched `pip` install of 2.1.0.0. When will this be updated?

**Comment by jimi-c** [comment]
@xenithorb this will be included in 2.1.1 (rc1 went out yesterday).

**Comment by psi-4ward** [comment]
Not fully working for me (2.1.1)

```
# task
...
  notify: update jenkins jdk config

# handler in same role:
- name: update jenkins jdk config
  include: jenkins-config.yml
  when: not inventory_hostname | search("slave") and not inventory_hostname | match(".*[1-9]$")
```

gives me: `ERROR! The requested handler 'update jenkins jdk config' was not found in any of the known handlers`

**Comment by elasticdog** [comment]
I get the exact same `The requested handler '[...]' was not found in any of the known handlers` error as @psi-4ward. If I change the handler to a non-include task, it finds the handler just fine.

**Comment by aabouzaid** [comment]
Handlers -even in a single playbook and no include- are not fired at all in 2.1.1.0!

> $ ansible --version
> ansible 2.1.1.0
>   config file = /etc/ansible/ansible.cfg
>   configured module search path = Default w/o overrides
> 
> $ lsb_release -a
> No LSB modules are available.
> Distributor ID: Ubuntu
> Description:    Ubuntu 14.04.4 LTS
> Release:    14.04
> Codename:   trusty

**Comment by jimi-c** [comment]
@AAbouZaid if that is the case, please open a new issue for it.

**Comment by aabouzaid** [comment]
@jimi-c After some debugging I found the problem was actually in the playbook itself (my problem not main one in this post).
Thanks a lot (y)

**Comment by mvdbeek** [comment]
As @elasticdog mentioned, this is still broken (though no traceback as in the initial issue description) in 2.1.1.0.

This is with the same steps @pstch has described.

```
(.venv)mariuss-MBP:demo_problem marius$ ansible --version
ansible 2.1.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
(.venv)mariuss-MBP:demo_problem marius$ ansible-playbook -c local playbook.yml

PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [demo_problem : Test task] ************************************************
ERROR! The requested handler 'Test' was not found in any of the known handlers
```

I'd appreciate it if someone (@jimi-c ?) could re-open the issue.

**Comment by jimi-c** [comment]
If the include is dynamic (using a loop or a variable in the include file name), you cannot target handler names inside it. As of 2.1, you can set config options in ansible.cfg (one each for regular tasks vs. handlers) or you can set `static: yes` on the include task as follows:

<pre>
handlers:
- include: "{{some_var}}.yml"
  static: yes
</pre>

which will force the evaluation of the variable `some_var` at playbook compilation time.

If you are still having problems after following the above suggestions, please let us know.

**Comment by elasticdog** [comment]
@jimi-c that suggestion doesn't apply in my case, the include is just a static filename (no loop or variable involved).

**Comment by mvdbeek** [comment]
yep, same here. The example posted by @pstch is a perfect minimal example to reproduce this bug.

**Comment by jimi-c** [comment]
Ahh, I see. This is actually the converse of the static situation - since in a static include the top level include "disappears" in a way. Reopening.

**Comment by jaccoh** [comment]
Another note... I think its the same bug though. The file is included but it seems the conditional is ignored. Or atleast in a weird way. The tasks are all skipped but they are included:

TASK [authdns-slave : set fact - master_host] **********************************
skipping: [lnx2347vm.internal.asp4all.nl] => (item=lnx2346vm.internal.asp4all.nl)

TASK [authdns-slave : set fact - master_ipv4_addr_internal] ********************
skipping: [lnx2347vm.internal.asp4all.nl]

TASK [authdns-slave : set fact - master_ipv4_addr_external] ********************
skipping: [lnx2347vm.internal.asp4all.nl]

TASK [authdns-slave : set fact - master_ipv6_addr_internal] ********************
skipping: [lnx2347vm.internal.asp4all.nl]

But those shouldnt have been running anyway:

```
- name: addzones part2
  include: addzones_part2.yml
  when: addzones
```

Expected behaviour:
In the output it should say that it skips "addzones part2" at include level.
addzones is False.

**Comment by mvdbeek** [comment]
> Ahh, I see. This is actually the converse of the static situation - since in a static include the top level include "disappears" in a way. Reopening.

That gave me the idea to try the dynamic include: that works fine (but is a bit nonsensical in this situation).

**Comment by jimi-c** [comment]
# Closing This Ticket

Hi!

We believe the above commit should resolve this problem for you. This will also be included in the next release.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!

**Comment by mvdbeek** [comment]
Thanks @jimi-c, this works now. Can we get this into the next 2.1 release?

**Comment by xenithorb** [comment]
I came across this again after the `2.1.1` fix as `the request handler [...] was not found` bug while doing vagrant tests. I noticed that despite the bug, usually running the playbook again (or in this case `vagrant provision`) would work. Has this been accounted for as some kind of race condition?

**Comment by jimi-c** [comment]
@mvdbeek I'm not 100% sure. The include stuff in devel has changed quite a bit due to some bug fixes and the performance improvements we made recently, so I'm not sure how cleanly this patch would apply to stable-2.1.

**Comment by jimi-c** [comment]
Yeah, it did not apply cleanly at all. If we want this in 2.1.2 (which we plan on releasing the first RC for next week), I'd have to rewrite the patch from scratch. I can look into doing that.

**Comment by mvdbeek** [comment]
I would very much appreciate it if an adapted patch would end up in 2.1.
I had a look too, [this](https://github.com/ansible/ansible/compare/stable-2.1...mvdbeek:fix_2.1_task_include?expand=1) seems to work, but I have no clue about the side effects of skipping `_block`. Thanks for your time @jimi-c.

**Comment by jimi-c** [comment]
@mvdbeek the above patch was merged into stable-2.1, and will be included in 2.1.2.

Thanks!

**Comment by mvdbeek** [comment]
Awesome, thanks a lot.

**Comment by x3medima17** [comment]
I used stable-2.1 and still got same error:
`ERROR! The requested handler 'restart rabbitmq' was not found in any of the known handlers`

A lot of tasks give this error, but each task gives it only once, i.e second time it goes well.

**Comment by ludovic-gasc** [comment]
I confirm the regression that @x3medima17 noticed, I've the same issue.

Could you reopen the issue ?

**Comment by jimi-c** [comment]
@x3medima17 / @GMLudo could you please share a playbook which reproduces the issue you're seeing?

**Comment by x3medima17** [comment]
@jimi-c here is the playbook

```
- name: Configure supervisor
  file:
    remote_src : True
    src : "{{ansible_env.HOME}}/worker/worker.conf"
    dest : /etc/supervisor/conf.d/worker.conf
    state : link
  notify:
    -restart supervisord
    - restart supervisor
  become : yes
```

And the error:
`ERROR! The requested handler '-restart supervisord - restart supervisor' was not found in any of the known handlers`

I have lost previous playbook

**Comment by jimi-c** [comment]
@x3medima17 that's a typo: you're missing a space in between the `-` and `restart supervisord`. Unfortunately, YAML is then reading this as a string (and, unhelpfully) reading the next line as a continuation of the previous line, thus the reason you see it reported as `-restart supervisord - restart supervisor`.

**Comment by x3medima17** [comment]
Strange, it should give syntax error. I have changed playbook

```
- name: Configure supervisor
  file:
    remote_src : True
    src : "{{ansible_env.HOME}}/worker/worker.conf"
    dest : /etc/supervisor/conf.d/worker.conf
    state : link
  notify:
    - restart supervisord
    - restart supervisor
  become : yes
```

And I got an error. it is very veird that error disappears when I run it again.

```
ERROR! The requested handler 'restart supervisord' was not found in any of the known handlers
```

**Comment by jimi-c** [comment]
@x3medima17 yeah, odd quirk of YAML there that it doesn't raise a syntax error. How is your handler defined? Is it inside an included file which is dynamic?

**Comment by x3medima17** [comment]
@jimi-c It is just a task from main file of the role: roles/worker/task/main.yml

**Comment by ajaybhatnagar** [comment]
I also get this error with latest ansible updated from ppa repo.  When I commented out notify step in task, it executed normally. Details of my case below: 

Error while running the playbook:

TASK [es_deploy/2.3.4 : Update ES Config] **************************************
ok: [10.236.133.188]
**_ERROR! The requested handler 'restart elasticsearch' was not found in any of the known handlers_**

**Version:**
ansible --version : ansible 2.1.1.0, on Ubuntu 14.04.

**Entries in ansible.cfg added based on threads in this discussion**:
task_includes_static = yes
handler_includes_static = yes
force_handlers = True

**Lines in role / task giving the error:**
- name: "Update ES Config"
    template: src=./templates/es_config.j2 dest={{es_conf_dir}}/elasticsearch.yml owner={{ES_USER}} group={{ES_GROUP}}
  
  **_#   notify: restart elasticsearch_**
    tags: [elasticsearch]
  - meta: flush_handlers

**Comment by mvdbeek** [comment]
@ajaybhatnagar 
we'd need to see your handler definition to be sure there isn't any other problem, but the fix for the issue discussed here will only be in 2.1.2.0, which hasn't been released yet.

**Comment by rdbisme** [comment]
**Ansible Version**: 2.1.2.0
**OS:** Archlinux + virtualenv (Python 2.7.12)

**main.yml**

```
---

- hosts: users 
  gather_facts: no
  vars:
          install_gnuradio_drivers: false

  tasks:

  - include: templates.yml
    become: yes

... Other tasks ...

  - include: handlers.yml
    become: yes

```

**handlers.yml**

```
- name: restart couchdb
  service: name=couchdb state=restarted
```

`ansible-playbook main.yml` returns the mentioned error only at first run. If I run it a second time everything works correctly.

**Comment by levequej** [comment]
Exact same problem here with ansible 2.2.1 on Centos7 , and same playbook was working fine with 2.2.0

as said by @x3medima17 and @rubendibattista
resmo > "A lot of tasks give this error, but each task gives it only once, i.e second time it goes well." 

example
```
ERROR! The requested handler 'postinstall syslog-ng' was not found in either the main handlers list nor in the listening handlers list
```
and the next time it goes on until it fails to the next handler

wondering where is the bug

**Comment by diego-treitos** [comment]
Also having this problem on `ansible 2.2.1.0`.

```
TASK [mysql : Add mysql configuration] *****************************************
ERROR! The requested handler 'restart mysql new size' was not found in either the main handlers list nor in the listening handlers list
```

`roles/mysl/handlers/main.yml`:
```
---
- name: restart mysql
  service: name=mysql state=restarted

- name: restart mysql new size
  include: restart_mysql_newsize.yml
```

`roles/mysql/tasks/main.yml`:
```
[...]

- name: Add mysql configuration
  template:
    src: my.cnf.j2
    dest: /etc/mysql/my.cnf
    owner: root
    group: root
    mode: 0640
  notify: restart mysql new size

[...]
```

**Comment by ludovic-gasc** [comment]
Hi,

Somebody could re-open this issue ?
The bug is still present.

Thank you.

**Comment by mickael-ange** [comment]
Hi,

Got the same issue with Ansible 2.3.0.0.
I worked around the issue by using a String and Jinja.

For instance: the following handler is not found, and I got `ERROR! The requested handler 'reload nginx' was not found in either the main handlers list nor in the listening handlers list`

`handlers/main.yml`

```
- name: reload nginx
  include: reload-nginx.yml
```

But this one is registered and called:

```
- name: reload nginx
#  include: reload-nginx.yml
  include: "reload-{{ 'nginx' }}.yml"
```
Noted that using `static: yes` gave the same `ERROR! The requested handler 'reload nginx' was not found in either the main handlers list nor in the listening handlers list`.

By the way I notice that the output of the tasks in `reload-nginx.yml` are not shown properly, instead the name of modules is displayed.

`tasks/reload-nginx.yml`
```
- include: check-syntax.yml

- name: Reloading Nginx
  become: yes
  service:
    name: nginx
    state: reloaded
  when: nginx_service_state != 'stopped'
```

`tasks/check-syntax.yml`

```
- name: Nginx Check syntax
  become: yes
  command: nginx -t
  ignore_errors: yes
  register: nginx_check_syntax

- debug:`
    var: nginx_check_syntax
  when: nginx_check_syntax | failed

- name: Fails if syntax check failed
  fail:
    msg: "Nginx configuration syntax check FAILED. See above output for details."
  when: nginx_check_syntax | failed
```
Output of the handler:
```
RUNNING HANDLER [nginx : reload nginx] ***************************************************
Friday 14 April 2017  00:40:15 +0800 (0:00:12.399)       0:00:27.449 ********** 
included: /home/mickael/git/emq/emq-devops/ansible/roles/nginx/tasks/reload-nginx.yml for allin

RUNNING HANDLER [nginx : include] *****************************************************************************************************
Friday 14 April 2017  00:40:15 +0800 (0:00:00.025)       0:00:27.475 ********** 
included: /home/mickael/git/emq/emq-devops/ansible/roles/nginx/tasks/check-syntax.yml for allin

RUNNING HANDLER [nginx : command] *****************************************************************************************************
Friday 14 April 2017  00:40:15 +0800 (0:00:00.027)       0:00:27.502 ********** 
changed: [allin]

RUNNING HANDLER [nginx : debug] *****************************************************************************************************
Friday 14 April 2017  00:40:15 +0800 (0:00:00.337)       0:00:27.840 ********** 
skipping: [allin]

RUNNING HANDLER [nginx : fail] *****************************************************************************************************
Friday 14 April 2017  00:40:15 +0800 (0:00:00.018)       0:00:27.858 ********** 
skipping: [allin]

RUNNING HANDLER [nginx : service] *****************************************************************************************************
Friday 14 April 2017  00:40:15 +0800 (0:00:00.019)       0:00:27.878 ********** 
changed: [allin]

```

Regards,
Mickael

**Comment by ludovic-gasc** [comment]
Hi,

Somebody could re-open this issue ?
The bug is still present.

Thank you.

**Comment by ludovic-gasc** [comment]
I confirm this bug is still present in the latest stable version.

We have rewritten our playbook to bypass that.

You might re-open this issue to not forget it's still broken.

**Comment by holmesb** [comment]
+1 I'm experiencing this in v2.4.  Please reopen this issue or advise if there's another one that is being used to track.

**Comment by jimi-c** [comment]
We do not typically re-open issues, as the root cause is probably not the same. If you're still seeing this, please open a new issue and we'll track it there.

Thanks!

**Comment by jimi-c** [comment]
And my apologies for not seeing this sooner.

**Comment by tunaranch** [comment]
This issue affects ansible 2.3.3.0 as well.

```
$ pip list
ansible (2.3.3.0)
asn1crypto (0.24.0)
bcrypt (3.1.4)
cffi (1.11.4)
cryptography (2.1.4)
enum34 (1.1.6)
idna (2.6)
ipaddress (1.0.19)
Jinja2 (2.10)
MarkupSafe (1.0)
paramiko (2.4.0)
pip (9.0.1)
pyasn1 (0.4.2)
pycparser (2.18)
pycrypto (2.6.1)
PyNaCl (1.2.1)
PyYAML (3.12)
setuptools (38.4.0)
six (1.11.0)
wheel (0.30.0)
```

The workaround described in https://github.com/ansible/ansible/issues/15915#issuecomment-293960640 gets around the problem.
